### PR TITLE
Adding of Join route on web part

### PR DESCRIPTION
**Pull request by iamgabrielsoft**
Currently there is no join route handler on the web part, for joining teams on Notex. The issue https://github.com/codex-team/notes.web/issues/183. This PR serves as a new feature for the implmentation

### CODE REVIEW

**Review by neSpecc — CHANGES_REQUESTED** [review]
Please, write a descriptive PR description. It should explain your solution and contain an information about changes you made.

Also, attach link correctly. Read this guide: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue

File: src/application/services/useNoteSettings.ts [hunk 1]
```diff
@@ -79,10 +92,21 @@ export default function (): UseNoteSettingsComposableState {
     }
   };
 
+  const joinNote = async (hash: string): Promise<TeamMember> => {
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:95]: use named function instead of anonymous

```suggestion
  async function joinNote(hash: string): Promise<TeamMember> {
```

Comment (neSpecc) [reply src/application/services/useNoteSettings.ts:95]: jsdoc is missed

File: src/application/services/useNoteSettings.ts [hunk 2]
```diff
@@ -79,10 +92,21 @@ export default function (): UseNoteSettingsComposableState {
...
+    if (!authService.repository.hasSession()) {
+      void router.push('/');
+    }
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:98]: leave a `@todo` to implement auth-guard on router level

File: src/domain/entities/TeamMember.ts [hunk 3]
```diff
@@ -0,0 +1,9 @@
+export interface TeamMember {
```
Comment (neSpecc) [inline src/domain/entities/TeamMember.ts:1]: docs are missed

File: src/domain/noteSettings.repository.interface.ts [hunk 4]
```diff
@@ -30,4 +31,11 @@ export default interface NoteSettingsRepositoryInterface {
    * @returns updated note settings
    */
   regenerateInvitationHash(id: NoteId): Promise<NoteSettings>;
+
+  /**
+   * Join note by hash
+   *
+   * @param hash - hash key
```
Comment (neSpecc) [inline src/domain/noteSettings.repository.interface.ts:38]:
```suggestion
   * @param hash - invitation hash
```

File: src/domain/noteSettings.service.ts [hunk 5]
```diff
@@ -86,4 +87,23 @@ export default class NoteService {
...
+   * @returns { TeamMember }
+   */
+  public async joinNoteTeam(hash: string): Promise<TeamMember> {
+    let result;
+
+    try {
+      result = this.noteSettingsRepository.joinNoteByInvitationHash(hash);
+    } catch (error) {
+      throw error;
+    }
```
Comment (neSpecc) [inline src/domain/noteSettings.service.ts:105]: this try-catch seems redundant

File: src/presentation/pages/Join.vue [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+<script lang="ts" setup>
+import { onMounted } from 'vue';
+import { useRoute } from 'vue-router';
+import useNoteSettings from '@/application/services/useNoteSettings';
+
+const route = useRoute();
+const { joinNote } = useNoteSettings();
+
+onMounted(async () => {
```
Comment (neSpecc) [inline src/presentation/pages/Join.vue:16]: maybe onBeforeRouteEnter?

Comment (iamgabrielsoft) [reply src/presentation/pages/Join.vue:16]: yes, but onBeforeRouteEnter is not available on composition api but on option api

except i need to import application service into routes.ts file

**Review by neSpecc — CHANGES_REQUESTED** [review]
Issue is not attached correctly.

File: src/domain/entities/TeamMember.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+   */
+  noteId: number;
+
+  /**
+   * team member userid
+   */
+  userId: string;
+
+  /**
+   * team member role
```
Comment (neSpecc) [inline src/domain/entities/TeamMember.ts:21]: Provided docs are useless. Please, write docs explaining the reason of each property exists.

File: src/presentation/pages/Join.vue (hunk 6 above)
Comment (neSpecc) [reply src/presentation/pages/Join.vue:16]: Read this guide, it can help
https://router.vuejs.org/guide/advanced/navigation-guards.html#The-Full-Navigation-Resolution-Flow

Probably, `beforeEnter` can be used here

File: src/presentation/pages/NoteSettings.vue [hunk 8]
```diff
@@ -56,8 +56,9 @@ const invitationLink = computed(
   () => `${import.meta.env.VITE_PRODUCTION_HOSTNAME}/join/${noteSettings.value?.invitationHash}`
 );
 
-load(props.id);
-
+onMounted(() => {
```
Comment (neSpecc) [inline src/presentation/pages/NoteSettings.vue:59]: describe why this change has been added

**Review by neSpecc — CHANGES_REQUESTED** [review]

**Comment by neSpecc** [comment]
> This PR provides a fix

it is not a fix but a new feature. Please, update a PR description

File: src/presentation/pages/Join.vue (hunk 6 above)
Comment (iamgabrielsoft) [reply src/presentation/pages/Join.vue:16]: its can but in routes.ts file

File: src/application/services/useNoteSettings.ts [hunk 9]
```diff
@@ -79,10 +92,28 @@ export default function (): UseNoteSettingsComposableState {
     }
   };
 
+  /**
+   * Join notex team when authenticated
+   *
+   * @todo - to be implemented as an auth guard at router level
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:98]: move a todo near auth checking statement

File: src/application/services/useNoteSettings.ts [hunk 10]
```diff
@@ -79,10 +92,28 @@ export default function (): UseNoteSettingsComposableState {
...
+   *
+   * @param hash - invitation hash
+   */
+  async function joinNote(hash: string): Promise<TeamMember> {
+    if (!authService.repository.hasSession()) {
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:103]: you should not access repository from application service

File: src/domain/auth.service.ts [hunk 11]
```diff
@@ -11,7 +11,7 @@ export default class AuthService {
   /**
    * Facade for accessing auth data
    */
-  private readonly repository: AuthRepository;
+  public readonly repository: AuthRepository;
```
Comment (neSpecc) [inline src/domain/auth.service.ts:14]: repo should be private for domain level

File: src/domain/noteSettings.service.ts [hunk 12]
```diff
@@ -86,4 +88,27 @@ export default class NoteService {
...
+
+    try {
+      result = await this.noteSettingsRepository.joinNoteByInvitationHash(hash);
+    } catch (error) {
+      if (error instanceof UnauthorizedError) {
+        throw new Error(`Authorization required to view this resource`);
+      }
+
+      throw error;
+    }
```
Comment (neSpecc) [inline src/domain/noteSettings.service.ts:110]: this catch seems useless. For what reason is it used?

File: src/presentation/pages/Join.vue (hunk 6 above)
Comment (neSpecc) [reply src/presentation/pages/Join.vue:16]: this comment is still actual. Please, read the whole article from Vue Router doc and select a corresponded hook

File: src/presentation/pages/NoteSettings.vue (hunk 8 above)
Comment (neSpecc) [reply src/presentation/pages/NoteSettings.vue:62]: still actual. Added doc is useless, it just repeats the code. It should describe why this code has been added?

**Review by neSpecc — CHANGES_REQUESTED** [review]
Most comments are still actual.
Issue is still not attached

File: src/domain/noteSettings.service.ts (hunk 12 above)
Comment (iamgabrielsoft) [reply src/domain/noteSettings.service.ts:110]: unauthorized access(403)

File: src/application/router/routes.ts [hunk 13]
```diff
@@ -57,6 +59,7 @@ const routes: RouteRecordRaw[] = [
     path: `/settings/`,
     component: Settings,
   },
+
```
Comment (neSpecc) [inline src/application/router/routes.ts:62]:
```suggestion
```

File: src/application/router/routes.ts [hunk 14]
```diff
@@ -65,6 +68,18 @@ const routes: RouteRecordRaw[] = [
...
+  {
+    name: 'join',
+    path: '/join/:id',
+    component: Join,
+    props: (route) => ({
+      id: String(route.params.id),
+    }),
+    beforeEnter: async (to) => {
+      await noteSettingsService.joinNoteTeam(to.params.id as string);
```
Comment (neSpecc) [inline src/application/router/routes.ts:80]: add error handling, please

File: src/application/services/useEditor.ts [hunk 15]
```diff
@@ -156,7 +156,7 @@ export function useEditor({ id, content, isReadOnly, onChange }: UseEditorParams
   }
 
   /**
-   * Reinitializes editor instance with new data
+   * Reinitialized editor instance with new data
```
Comment (neSpecc) [inline src/application/services/useEditor.ts:159]: seems like unexpected change

File: src/application/services/useNoteSettings.ts [hunk 16]
```diff
@@ -79,10 +91,26 @@ export default function (): UseNoteSettingsComposableState {
     }
   };
 
+  /**
+   * Join notex team when authenticated
+   *
+   * @param hash - invitation hash
+   */
+  async function joinNote(hash: string): Promise<TeamMember> {
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:99]: is it used?

File: src/presentation/pages/Join.vue
Comment (neSpecc) [inline src/presentation/pages/Join.vue:1]: seems useless

File: src/presentation/pages/NoteSettings.vue (hunk 8 above)
Comment (neSpecc) [reply src/presentation/pages/NoteSettings.vue:62]: still ignored

File: src/domain/auth.service.ts [hunk 17]
```diff
@@ -74,6 +75,13 @@ export default class AuthService {
     this.eventBus.dispatchEvent(new AuthLogoutEvent());
   }
 
+  /**
+   * checker for authorized user
```
Comment (neSpecc) [inline src/domain/auth.service.ts:79]:
```suggestion
   * Checks whether user is authorized or not
```

File: src/domain/auth.service.ts [hunk 18]
```diff
@@ -50,6 +50,7 @@ export default class AuthService {
   /**
    * Called after oauth to accept session
    *
+   * @todo - to be implemented as an auth guard at router level
```
Comment (neSpecc) [inline src/domain/auth.service.ts:53]: wrong place for "todo". Add it to a related code fragment

**Review by neSpecc — CHANGES_REQUESTED** [review]

File: src/application/services/useNoteSettings.ts [hunk 19]
```diff
@@ -79,10 +91,26 @@ export default function (): UseNoteSettingsComposableState {
...
+    if (!authService.isAuthorized()) {
+      void router.push('/');
+    }
+
+    teamMember.value = await noteSettingsService.joinNoteTeam(hash);
```
Comment (elizachi) [inline src/application/services/useNoteSettings.ts:104]: User is not joining the team, you need to fix that

File: src/domain/entities/TeamMember.ts [hunk 20]
```diff
@@ -0,0 +1,24 @@
...
+
+  /**
+   * User(userId) of the user associated with the team
+   */
+  userId: string;
+
+  /**
+   * role of the team member e.g either READ(0) or WRITE(1) role
+   */
+  role: string;
```
Comment (elizachi) [inline src/domain/entities/TeamMember.ts:23]: Maybe add enums like in api? It can helps with data validation

**Review by elizachi — COMMENTED** [review]
Please update project to latest version

**Review by elizachi — CHANGES_REQUESTED** [review]
I also think it's worth renaming the branch name

**Comment by iamgabrielsoft** [comment]
> I also think it's worth renaming the branch name

I prefer using one branch for things similar, I would've to rename the branch to something more generic

File: src/application/router/routes.ts [hunk 21]
```diff
@@ -65,6 +67,24 @@ const routes: RouteRecordRaw[] = [
...
+    name: 'join',
+    path: '/join/:id',
+    component: Join,
+    props: route => ({
+      id: String(route.params.id),
+    }),
+    beforeEnter: async (to, from, next) => {
+      const noteSettings = await useNoteSettings().joinNote(to.params.id as string);
+
+      if (noteSettings?.id != null) {
```
Comment (neSpecc) [inline src/application/router/routes.ts:81]: you need to handle joining errors and show user what happened and what he should need to do

File: src/application/services/useNoteSettings.ts [hunk 22]
```diff
@@ -126,12 +138,33 @@ export default function (): UseNoteSettingsComposableState {
     });
   };
 
+  async function joinNote(hash: string): Promise<TeamMember> {
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:141]: docs missed

File: src/application/services/useNoteSettings.ts [hunk 23]
```diff
@@ -126,12 +138,33 @@ export default function (): UseNoteSettingsComposableState {
...
+    if (!authService.isAuthorized()) {
+      void router.push('/');
+      void router.push('/');
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:144]: why two times?

File: src/application/services/useNoteSettings.ts [hunk 24]
```diff
@@ -126,12 +138,33 @@ export default function (): UseNoteSettingsComposableState {
...
+
+      if (route.path === '/') {
+        void useAuth().showGoogleAuthPopup();
+        // setTimeout(() => {
+        //
+        // //router.push({ name: 'join', params: { id: hash }})
+        // }, 4000)
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:151]: why this comments is needed for?

File: src/application/services/useNoteSettings.ts [hunk 25]
```diff
@@ -126,12 +138,33 @@ export default function (): UseNoteSettingsComposableState {
...
+      }
+    }
+
+    teamMember.value = await noteSettingsService.joinNoteTeam(hash);
+    teamMember.value = await noteSettingsService.joinNoteTeam(hash);
```
Comment (neSpecc) [inline src/application/services/useNoteSettings.ts:156]: duplicate

**Review by neSpecc — CHANGES_REQUESTED** [review]
Please, review the PR by yourself and fix all the issued you'll find. Right now it seems unready.

File: src/application/router/routes.ts (hunk 21 above)
Comment (iamgabrielsoft) [reply src/application/router/routes.ts:81]: currently when a user isn't authenticated, he's signed or using google oauth popu. To display message i think i need to keep track of state

Please line 147, useNoteSettings.ts file

Comment (iamgabrielsoft) [reply src/application/router/routes.ts:81]: currently when a user isn't authenticated, he's signed or using google oauth popu. To display message i think i need to keep track of state

Please line 147, useNoteSettings.ts file

Comment (iamgabrielsoft) [reply src/application/router/routes.ts:81]: > you need to handle joining errors and show user what happened and what he should need to do

currently when a user isn't authenticated, he's signed or using google oauth popu. To display message i think i need to keep track of state

Please line 147, useNoteSettings.ts file

Comment (iamgabrielsoft) [reply src/application/router/routes.ts:81]: In that case, i don't think handling the error on route file would help me to display 'what the user needs to do' rather on join.vue file would. Since i can't track state as its happens

File: src/presentation/pages/Join.vue [hunk 26]
```diff
@@ -0,0 +1,30 @@
...
+  useNoteSettings().joinNote(route.params.id as string)
+    .then(() => {})
+    .catch(err => message.value = err.message);
+}
+
+checkJoinEligibility();
+
+useHead({
+  title: t('join.title'),
+});
```
Comment (neSpecc) [inline src/presentation/pages/Join.vue:29]: it is a technical route, ideally user should now see it, so the title is redundant

File: src/presentation/pages/Join.vue [hunk 27]
```diff
@@ -0,0 +1,30 @@
+<template>
+  <h1>Team</h1>
+  <Section :title="t(message)" />
```
Comment (neSpecc) [inline src/presentation/pages/Join.vue:3]: we should not display API errors to the user directly.

File: src/presentation/pages/NoteSettings.vue [hunk 28]
```diff
@@ -101,7 +105,7 @@ const isPublic = computed(() => {
  * Change isPublic property
  */
 async function changeAccess() {
-  updateIsPublic(props.id, !noteSettings.value!.isPublic);
+  await updateIsPublic(props.id, !noteSettings.value);
```
Comment (neSpecc) [inline src/presentation/pages/NoteSettings.vue:108]: irrelevant change

File: src/presentation/pages/NoteSettings.vue [hunk 29]
```diff
@@ -70,23 +70,27 @@ const invitationLink = computed(
   () => `${import.meta.env.VITE_PRODUCTION_HOSTNAME}/join/${noteSettings.value?.invitationHash}`
 );
 
+/**
+ * Button click handler
+ */
```
Comment (neSpecc) [inline src/presentation/pages/NoteSettings.vue:75]: where this doc came from?
